### DISCUSSION TITLE
Consider supporting multiple exports from a single file

### DISCUSSION THREAD

**Discussion by josetr** [vikejs/vike#1354] [Discussion]
https://github.com/vikejs/vike/blob/a862b2a0c94cfc56f63c817bf633c2202fe09e6a/vike/shared/page-configs/assertExports.ts#L81

Looks like there is an exception for '.md' and '.mdx' files.

**Comment by brillout** [comment]
I disagree, but feel free to elaborate on why you believe/need this.

**Comment by josetr** [comment]
Vike claims that it allows you to keep "Architectural control" and "Build Your Own Framework" so it should be flexible here, otherwise it's impossible to use it to create a framework that provides a similar DX such as Remix or Next.js, where multiple exports is allowed.

**Reply by brillout** [reply]
I don't think this kind of control to be important. I believe what matters is architecural control, in order to  be blocked as less as possible.

**Reply by josetr** [reply]
I believe it is, and you also show a good reason at https://github.com/vikejs/vike/issues/1056.

The Svelte framework also suffers from the 1 component per file limitation, and it's not hard to find people complain about it, and even claim that they don't use Svelte due to this limitation.

Most people are using Next.js or Remix these days, and already got used to having the ability to export multiple "hooks" from the page file. If you want to restrict them, it would be nice if you could at least give a good reason to do so.

It's not like supporting it's complicated, since it already kinda works, it just throws warnings.

**Reply by brillout** [reply]
> 1 component per file 

That's a *much* harder restriction. I actually agree on that one and it's something that turns me off about Svelte and Vue in a non-negligible manner.

Whereas 1 `+` file per Vike hook is, I believe, negligible.

Instead of defining `+` files, you can also define multiple hooks in one file then import them in your `+config.h.js`.

> It's not like supporting it's complicated

It actually is.

**Reply by josetr** [reply]
I guess I'm lost at the part where you claim it's complicated, since like I said, it already works.

1 file per Vike hook makes sense most of the time, but not always. If you are starting from scratch, it's an annoyance, but if you are trying to migrate from another framework, with a large codebase, it's a problem. Specially when are unsure about trying a new framework, and you may want to have an easy path to go back to the previous framework.

At the end of the day, this is like the Tab vs Space problem, where there is no right answer, and it's down to preferences, so both options should be supported, unless there is a good reason not to do so, which I haven't heard so far.
